### PR TITLE
Better ClientInfo parser

### PR DESCRIPTION
**Pull request by alxdarksage**
Can parse a much wider range of combinations that are, nevertheless, intelligible as user agent strings, e.g. "/7 (Android/10) /5" should still be parseable, and now it is.

### CODE REVIEW

File: test/org/sagebionetworks/bridge/models/ClientInfoTest.java [hunk 1]
```diff
@@ -15,229 +10,152 @@
...
+        assertEquals(sdkVersion, info.getSdkVersion());
     }
     
     @Test
-    public void deprecatedOperatingSystemParsed() {
-        ClientInfo info = ClientInfo.parseUserAgentString(VALID_LONG_UA_3_DEPRECATED);
-        assertEquals(ANDROID, info.getOsName());
+    public void returnsUknownClient() {
+        ClientInfo info = ClientInfo.parseUserAgentString("/7.1.1.");
+        assertSame(ClientInfo.UNKNOWN_CLIENT, info);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/ClientInfoTest.java]: Should we have more test cases for this? Things that our parse throws out as invalid:

* Strings that start with ; or )
* Double open parentheses.
* ;; or ;(
* )) or )( or );
* /; or //( or ///
* ()/()

File: app/org/sagebionetworks/bridge/models/ClientInfo.java [hunk 2]
```diff
@@ -259,74 +241,188 @@ public static ClientInfo fromUserAgentCache(String userAgent) {
...
+    private static void parseAppStanza(ClientInfo.Builder builder, String stanza) {
+        if (StringUtils.isBlank(stanza)) {
+            return;
+        }
+        String[] components = stanza.split("/");
+        if (components.length == 2) {
+            if (!components[0].matches(DIGITS_REGEXP)) {
+                builder.withAppName(parseString(components[0]));    
+            }
+            builder.withAppVersion(parseInteger(components[1])).build();
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/ClientInfo.java]: Minor: Stray build? (And in parseOsStanza() and parseSdkStanza().)

File: app/org/sagebionetworks/bridge/models/ClientInfo.java [hunk 3]
```diff
@@ -259,74 +241,188 @@ public static ClientInfo fromUserAgentCache(String userAgent) {
...
+    
+    private static void parseAppStanza(ClientInfo.Builder builder, String stanza) {
+        if (StringUtils.isBlank(stanza)) {
+            return;
+        }
+        String[] components = stanza.split("/");
+        if (components.length == 2) {
+            if (!components[0].matches(DIGITS_REGEXP)) {
+                builder.withAppName(parseString(components[0]));    
+            }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/ClientInfo.java:358]: It looks like this code is to filter out appNames/Versions like 12/34, which parses to (no app name) version 34. Is this correct? Should we add a unit test case for this?

File: test/org/sagebionetworks/bridge/models/ClientInfoTest.java [hunk 4]
```diff
@@ -15,229 +10,157 @@
...
-        assertEquals(IOS, info.getOsName());
-        assertEquals("9.1", info.getOsVersion());
-        assertEquals("BridgeSDK", info.getSdkName());
-        assertEquals(4, info.getSdkVersion().intValue());
+        // Or the right side stanza. Verify all combinations of that stanza as well.
+        assertClientInfo("(deviceName; osName/4.0.0) sdkName", null, null, "deviceName", "osName", "4.0.0", "sdkName", null);
+        assertClientInfo("(deviceName; osName/4.0.0) /5", null, null, "deviceName", "osName", "4.0.0", null, 5);
+        assertClientInfo("(deviceName; osName/4.0.0) sdkName/5", null, null, "deviceName", "osName", "4.0.0", "sdkName", 5);
+        assertClientInfo("appName/1.0 (deviceName; osName/4.0.0) sdkName/5.0.0", "appName", null, "deviceName",
+                "osName", "4.0.0", "sdkName", null);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/ClientInfoTest.java:52]: Test case where sdk stanza is just a number (without a slash)?

File: test/org/sagebionetworks/bridge/models/ClientInfoTest.java [hunk 5]
```diff
@@ -15,229 +10,157 @@
...
-        assertEquals(ANDROID, info.getOsName());
-        assertEquals("14", info.getOsVersion());
-        assertEquals("BridgeJavaSDK", info.getSdkName());
-        assertEquals(10, info.getSdkVersion().intValue());
-    }
-    
-    @Test
-    public void cacheWorks() {
-        ClientInfo info1 = ClientInfo.fromUserAgentCache(VALID_LONG_UA_1);
-        ClientInfo info2 = ClientInfo.fromUserAgentCache(VALID_LONG_UA_1);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/ClientInfoTest.java]: Do we need the fromUserAgentCache() tests?

File: test/org/sagebionetworks/bridge/models/ClientInfoTest.java [hunk 6]
```diff
@@ -15,229 +10,157 @@
...
-    @Test
-    public void cacheWorksWithBadValues() {
-        assertSame(ClientInfo.UNKNOWN_CLIENT, ClientInfo.fromUserAgentCache(null));
-        assertSame(ClientInfo.UNKNOWN_CLIENT, ClientInfo.fromUserAgentCache("   \n"));
-    }
-    
-    @Test
-    public void testIsSupportedAppVersion_GreaterThanSucceeds() {
-        ClientInfo info = ClientInfo.parseUserAgentString(VALID_LONG_UA_1);
-        assertTrue(info.isSupportedAppVersion(25));
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/ClientInfoTest.java]: Do we need isSupportedAppVersion() tests?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: test/org/sagebionetworks/bridge/models/ClientInfoTest.java (hunk 1 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/ClientInfoTest.java]: Added all of these. All will result in UNKNOWN_CLIENT

File: app/org/sagebionetworks/bridge/models/ClientInfo.java (hunk 2 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/ClientInfo.java]: Yes! Removed.

File: app/org/sagebionetworks/bridge/models/ClientInfo.java (hunk 3 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/ClientInfo.java:358]: Yes. Added a couple of test and caught the fact that we're not doing this for the SDK stanza, so added the check there as well, so for example, "1/2 3/4" only parses appVersion = 2 and sdkVersion = 4.

File: test/org/sagebionetworks/bridge/models/ClientInfoTest.java (hunk 4 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/ClientInfoTest.java:52]: Added.

File: test/org/sagebionetworks/bridge/models/ClientInfoTest.java (hunk 6 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/ClientInfoTest.java]: Yes I was overly zealous in deleting tests of methods besides the parsing method itself. Restored these.

File: test/org/sagebionetworks/bridge/models/ClientInfoTest.java (hunk 5 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/ClientInfoTest.java]: Yes, added them back. I deleted too many tests

File: app/org/sagebionetworks/bridge/models/ClientInfo.java [hunk 7]
```diff
@@ -259,74 +243,190 @@ public static ClientInfo fromUserAgentCache(String userAgent) {
...
+            components = new String[] {
+                stanza.substring(0, stanza.lastIndexOf(" ")),
+                stanza.substring(stanza.lastIndexOf(" ")),
+            };
+        } else {
+            components = stanza.split("/");    
+        }
+        if (components.length == 2) {
+            builder.withOsName(parseString(components[0]));
+            builder.withOsVersion(parseString(components[1])).build();
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/ClientInfo.java]: Stray .build()

File: app/org/sagebionetworks/bridge/models/ClientInfo.java [hunk 8]
```diff
@@ -259,74 +243,190 @@ public static ClientInfo fromUserAgentCache(String userAgent) {
...
+        }
+    }
+    
+    private static void parseSdkStanza(ClientInfo.Builder builder, String stanza) {
+        String[] sdkComponents = stanza.split("/");
+        if (sdkComponents.length == 2) {
+            if (!sdkComponents[0].matches(DIGITS_REGEXP)) {
+                builder.withSdkName(parseString(sdkComponents[0]));    
+            }
+            builder.withSdkVersion(parseInteger(sdkComponents[1])).build();
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/ClientInfo.java]: Stray .build()

**Review by DwayneJengSage — APPROVED** [review]
A few minor comments. Otherwise, looks good.

**Pull request merged**